File: agent/history.mjs
import { existsSync, readFileSync, readdirSync } from "node:fs";
import { fileURLToPath } from "node:url";
import path from "node:path";

const HERE = path.dirname(fileURLToPath(import.meta.url));
const ROOT = path.resolve(HERE, "..");
const RESULTS = path.join(ROOT, "test-results");

function readJson(file) {
  try { return JSON.parse(readFileSync(file, "utf8")); }
  catch { return null; }
}

function manifestsBelow(root) {
  if (!existsSync(root)) return [];
  const files = [];
  const pending = [root];
  while (pending.length) {
    const directory = pending.pop();
    for (const entry of readdirSync(directory, { withFileTypes: true })) {
      const file = path.join(directory, entry.name);
      if (entry.isDirectory()) pending.push(file);
      else if (entry.isFile() && entry.name === "run-manifest.json") files.push(file);
    }
  }
  return files;
}

function percentile(values, fraction) {
  if (!values.length) return null;
  const ordered = [...values].sort((left, right) => left - right);
  return ordered[Math.min(ordered.length - 1, Math.ceil(ordered.length * fraction) - 1)];
}

function normalizedStatus(manifest) {
  if (manifest.status === "passed") return "passed";
  if (manifest.status === "blocked") return "blocked";
  if (manifest.status === "canceled") return "canceled";
  if (manifest.status === "executed") return "passed";
  if (manifest.status === "failed") return "failed";
  return manifest.completedAt ? "failed" : "incomplete";
}

function runRecord(manifestPath) {
  const manifest = readJson(manifestPath);
  if (!manifest) return null;
  const directory = path.dirname(manifestPath);
  const analysis = readJson(manifest.analysisPath || path.join(directory, "analysis.json"));
  const report = readJson(manifest.paths?.reportPath || path.join(directory, "report.json"));
  const failureText = (analysis?.failures || report?.failures || [])
    .map((failure) => failure.error || failure.message || "")
    .join("\n");
  const failureClass = /executable doesn't exist|driver.*not installed|toolchain|connection refused|econnrefused/i.test(failureText)
    ? "infrastructure"
    : "product";
  const testsByTitle = new Map();
  for (const test of analysis?.tests || report?.tests || []) {
    testsByTitle.set(test.title, {
      title: test.title,
      status: test.status || (test.passed ? "passed" : "failed"),
      durationMs: Number(test.durationMs ?? test.duration ?? 0),
    });
  }
  const tests = [...testsByTitle.values()];
  return {
    runId: manifest.runId,
    appId: manifest.appId,
    kind: manifest.kind || "adhoc",
    target: manifest.target,
    spec: manifest.spec || null,
    status: normalizedStatus(manifest),
    startedAt: manifest.startedAt,
    completedAt: manifest.completedAt || null,
    durationMs: Number(manifest.durationMs || 0),
    harness: manifest.harness || null,
    source: manifest.source || null,
    build: manifest.build || null,
    journeys: manifest.appManifest?.journeys || [],
    failureClass: normalizedStatus(manifest) === "failed" ? failureClass : null,
    device: manifest.device || null,
    conditions: manifest.conditions || {},
    evidence: manifest.evidence || null,
    artifacts: manifest.artifacts || [],
    protection: manifest.protection || null,
    manifestPath,
    analysisPath: manifest.analysisPath || null,
    tests,
  };
}

function testHistory(runs) {
  const byTitle = new Map();
  for (let runIndex = runs.length - 1; runIndex >= 0; runIndex -= 1) {
    const run = runs[runIndex];
    if (run.failureClass === "infrastructure") continue;
    for (const test of run.tests) {
      const observations = byTitle.get(test.title) || [];
      observations.push({ runId: run.runId, at: run.startedAt, status: test.status, durationMs: test.durationMs });
      byTitle.set(test.title, observations);
    }
  }
  return [...byTitle.entries()].map(([title, observations]) => {
    const passed = observations.filter((item) => item.status === "passed").length;
    let transitions = 0;
    for (let index = 1; index < observations.length; index += 1) {
      if (observations[index].status !== observations[index - 1].status) transitions += 1;
    }
    const durations = observations.filter((item) => item.status === "passed").map((item) => item.durationMs);
    return {
      title,
      observations: observations.length,
      passed,
      failed: observations.length - passed,
      passRate: observations.length ? passed / observations.length : null,
      transitions,
      flaky: transitions > 0 && passed > 0 && passed < observations.length,
      latest: observations.at(-1) || null,
      duration: {
        p50Ms: percentile(durations, 0.5),
        p95Ms: percentile(durations, 0.95),
        maxMs: durations.length ? Math.max(...durations) : null,
      },
    };
  }).sort((left, right) => left.title.localeCompare(right.title));
}

function journeyHistory(runs) {
  const names = new Set(runs.flatMap((run) => run.journeys));
  return [...names].sort().map((journey) => {
    const relevant = runs.filter((run) => run.journeys.includes(journey));
    const passed = relevant.filter((run) => run.status === "passed").length;
    const productFailures = relevant.filter((run) => run.status === "failed" && run.failureClass !== "infrastructure").length;
    return {
      journey,
      runs: relevant.length,
      passed,
      failed: relevant.filter((run) => run.status === "failed").length,
      productFailures,
      infrastructureFailures: relevant.filter((run) => run.status === "failed" && run.failureClass === "infrastructure").length,
      blocked: relevant.filter((run) => run.status === "blocked").length,
      canceled: relevant.filter((run) => run.status === "canceled").length,
      passRate: passed + productFailures ? passed / (passed + productFailures) : null,
      latestRunId: relevant[0]?.runId || null,
    };
  });
}

function performanceTrend(runs) {
  const passed = runs.filter((run) => run.status === "passed" && run.durationMs > 0);
  const latest = passed[0] || null;
  const baselineValues = passed.slice(1, 11).map((run) => run.durationMs);
  const baselineMs = percentile(baselineValues, 0.5);
  const ratio = latest && baselineMs ? latest.durationMs / baselineMs : null;
  return {
    latestMs: latest?.durationMs || null,
    baselineMedianMs: baselineMs,
    ratio,
    regression: ratio !== null && baselineMs >= 500 && ratio >= 1.2,
  };
}

export function runHistory({ appId = "probierz", target, limit = 50 } = {}) {
  const root = path.join(RESULTS, appId, ...(target ? [target.replace(/:/g, "-")] : []));
  const runs = manifestsBelow(root)
    .map(runRecord)
    .filter(Boolean)
    .filter((run) => !target || run.target === target)
    .sort((left, right) => String(right.startedAt).localeCompare(String(left.startedAt)))
    .slice(0, Math.max(1, Number(limit) || 50));
  const passed = runs.filter((run) => run.status === "passed").length;
  const failed = runs.filter((run) => run.status === "failed").length;
  const infrastructureFailures = runs.filter((run) => run.status === "failed" && run.failureClass === "infrastructure").length;
  const productFailures = failed - infrastructureFailures;
  const blocked = runs.filter((run) => run.status === "blocked").length;
  const canceled = runs.filter((run) => run.status === "canceled").length;
  const tests = testHistory(runs);
  const performance = performanceTrend(runs);
  return {
    schemaVersion: 2,
    appId,
    target: target || null,
    generatedAt: new Date().toISOString(),
    summary: {
      runs: runs.length,
      passed,
      failed,
      blocked,
      canceled,
      productFailures,
      infrastructureFailures,
      passRate: passed + productFailures ? passed / (passed + productFailures) : null,
      flakyTests: tests.filter((test) => test.flaky).length,
      latestRunId: runs[0]?.runId || null,
      lastGreenRunId: runs.find((run) => run.status === "passed")?.runId || null,
      performanceRegression: performance.regression,
    },
    performance,
    journeys: journeyHistory(runs),
    tests,
    runs,
  };
}
export function getRun(appId, runId) {
  const run = manifestsBelow(path.join(RESULTS, appId))
    .map(runRecord)
    .find((candidate) => candidate?.runId === runId);
  if (!run) throw new Error(`run not found for ${appId}: ${runId}`);
  return run;
}

function artifactIndex(run) {
  return new Map((run.artifacts || []).map((artifact) => [artifact.file, artifact]));
}

export function compareRuns({ appId = "probierz", leftRunId, rightRunId } = {}) {
  const left = getRun(appId, leftRunId);
  const right = getRun(appId, rightRunId);
  const leftTests = new Map(left.tests.map((test) => [test.title, test]));
  const rightTests = new Map(right.tests.map((test) => [test.title, test]));
  const testNames = [...new Set([...leftTests.keys(), ...rightTests.keys()])].sort();
  const testChanges = testNames.flatMap((title) => {
    const before = leftTests.get(title) || null;
    const after = rightTests.get(title) || null;
    if (!before) return [{ title, change: "added", before, after }];
    if (!after) return [{ title, change: "removed", before, after }];
    if (before.status !== after.status || before.durationMs !== after.durationMs) {
      return [{
        title,
        change: before.status === after.status ? "duration" : "status",
        before,
        after,
        durationDeltaMs: after.durationMs - before.durationMs,
      }];
    }
    return [];
  });
  const leftArtifacts = artifactIndex(left);
  const rightArtifacts = artifactIndex(right);
  const artifactNames = [...new Set([...leftArtifacts.keys(), ...rightArtifacts.keys()])].sort();
  const artifactChanges = artifactNames.flatMap((file) => {
    const before = leftArtifacts.get(file) || null;
    const after = rightArtifacts.get(file) || null;
    if (!before) return [{ file, change: "added", before, after }];
    if (!after) return [{ file, change: "removed", before, after }];
    if (before.sha256 !== after.sha256 || before.bytes !== after.bytes) {
      return [{ file, change: "content", before, after }];
    }
    return [];
  });
  const durationDeltaMs = right.durationMs - left.durationMs;
  const durationRatio = left.durationMs > 0 ? right.durationMs / left.durationMs : null;
  const newlyFailing = testChanges
    .filter((change) => change.after?.status === "failed" && change.before?.status !== "failed")
    .map((change) => change.title);
  return {
    schemaVersion: 2,
    appId,
    left: {
      runId: left.runId,
      status: left.status,
      harness: left.harness,
      source: left.source,
      build: left.build,
      durationMs: left.durationMs,
      evidence: left.evidence,
    },
    right: {
      runId: right.runId,
      status: right.status,
      harness: right.harness,
      source: right.source,
      build: right.build,
      durationMs: right.durationMs,
      evidence: right.evidence,
    },
    verdict: {
      statusChanged: left.status !== right.status,
      regression: right.status === "failed" && left.status === "passed",
      newlyFailing,
      durationRegression: durationRatio !== null && left.durationMs >= 500 && durationRatio >= 1.2,
    },
    duration: { deltaMs: durationDeltaMs, ratio: durationRatio },
    tests: { changed: testChanges.length, changes: testChanges },
    artifacts: { changed: artifactChanges.length, changes: artifactChanges },
  };
}

export function lastGreen({ appId = "probierz", target, journey } = {}) {
  const history = runHistory({ appId, target, limit: 1000 });
  const run = history.runs.find((candidate) => candidate.status === "passed"
    && (!journey || candidate.journeys.includes(journey)));
  return {
    schemaVersion: 2,
    appId,
    target: target || null,
    journey: journey || null,
    run: run || null,
  };
}

File: probierz-rs/src/status/history.rs
//! Run history: per-test and per-journey history, the performance trend, and the history answer.

use super::*;

pub(super) fn percentile(mut values: Vec<f64>, fraction: f64) -> Option<f64> {
    if values.is_empty() {
        return None;
    }
    values.sort_by(|left, right| left.partial_cmp(right).unwrap_or(Ordering::Equal));
    let index = ((values.len() as f64 * fraction).ceil() as usize).saturating_sub(1);
    values.get(index.min(values.len() - 1)).copied()
}

pub(super) fn test_history(runs: &[Value]) -> Vec<Value> {
    let mut order = Vec::new();
    let mut by_title: HashMap<String, Vec<Value>> = HashMap::new();
    for run in runs.iter().rev() {
        if run.get("failureClass").and_then(Value::as_str) == Some("infrastructure") {
            continue;
        }
        for test in run
            .get("tests")
            .and_then(Value::as_array)
            .into_iter()
            .flatten()
        {
            let Some(title) = test.get("title").and_then(Value::as_str) else {
                continue;
            };
            if !by_title.contains_key(title) {
                order.push(title.to_string());
            }
            by_title.entry(title.to_string()).or_default().push(json!({
                "runId": value_or(run.get("runId"), Value::Null),
                "at": value_or(run.get("startedAt"), Value::Null),
                "status": value_or(test.get("status"), Value::Null),
                "durationMs": value_or(test.get("durationMs"), json!(0)),
            }));
        }
    }
    order.sort();
    order
        .into_iter()
        .map(|title| {
            let observations = by_title.remove(&title).unwrap_or_default();
            let passed = observations
                .iter()
                .filter(|item| item.get("status").and_then(Value::as_str) == Some("passed"))
                .count();
            let transitions = observations
                .windows(2)
                .filter(|window| window[0].get("status") != window[1].get("status"))
                .count();
            let durations = observations
                .iter()
                .filter(|item| item.get("status").and_then(Value::as_str) == Some("passed"))
                .map(|item| number(item.get("durationMs")))
                .collect::<Vec<_>>();
            let p50 = percentile(durations.clone(), 0.5).map(json_number).unwrap_or(Value::Null);
            let p95 = percentile(durations.clone(), 0.95).map(json_number).unwrap_or(Value::Null);
            let max = durations
                .iter()
                .copied()
                .reduce(f64::max)
                .map(json_number)
                .unwrap_or(Value::Null);
            let total = observations.len();
            json!({
                "title": title,
                "observations": total,
                "passed": passed,
                "failed": total - passed,
                "passRate": if total > 0 { json_number(passed as f64 / total as f64) } else { Value::Null },
                "transitions": transitions,
                "flaky": transitions > 0 && passed > 0 && passed < total,
                "latest": observations.last().cloned().unwrap_or(Value::Null),
                "duration": { "p50Ms": p50, "p95Ms": p95, "maxMs": max },
            })
        })
        .collect()
}

pub(super) fn journey_history(runs: &[Value]) -> Vec<Value> {
    let names = runs
        .iter()
        .flat_map(|run| {
            run.get("journeys")
                .and_then(Value::as_array)
                .into_iter()
                .flatten()
        })
        .filter_map(Value::as_str)
        .map(str::to_string)
        .collect::<BTreeSet<_>>();
    names
        .into_iter()
        .map(|journey| {
            let relevant = runs
                .iter()
                .filter(|run| {
                    run.get("journeys")
                        .and_then(Value::as_array)
                        .is_some_and(|items| items.iter().any(|item| item.as_str() == Some(&journey)))
                })
                .collect::<Vec<_>>();
            let count_status = |wanted: &str| {
                relevant
                    .iter()
                    .filter(|run| run.get("status").and_then(Value::as_str) == Some(wanted))
                    .count()
            };
            let passed = count_status("passed");
            let product_failures = relevant
                .iter()
                .filter(|run| {
                    run.get("status").and_then(Value::as_str) == Some("failed")
                        && run.get("failureClass").and_then(Value::as_str) != Some("infrastructure")
                })
                .count();
            let infrastructure_failures = relevant
                .iter()
                .filter(|run| {
                    run.get("status").and_then(Value::as_str) == Some("failed")
                        && run.get("failureClass").and_then(Value::as_str) == Some("infrastructure")
                })
                .count();
            json!({
                "journey": journey,
                "runs": relevant.len(),
                "passed": passed,
                "failed": count_status("failed"),
                "productFailures": product_failures,
                "infrastructureFailures": infrastructure_failures,
                "blocked": count_status("blocked"),
                "canceled": count_status("canceled"),
                "passRate": if passed + product_failures > 0 {
                    json_number(passed as f64 / (passed + product_failures) as f64)
                } else { Value::Null },
                "latestRunId": relevant.first().and_then(|run| run.get("runId")).cloned().unwrap_or(Value::Null),
            })
        })
        .collect()
}

pub(super) fn performance_trend(runs: &[Value]) -> Value {
    let passed = runs
        .iter()
        .filter(|run| {
            run.get("status").and_then(Value::as_str) == Some("passed")
                && number(run.get("durationMs")) > 0.0
        })
        .collect::<Vec<_>>();
    let latest_ms = passed.first().map(|run| number(run.get("durationMs")));
    let baseline = percentile(
        passed
            .iter()
            .skip(1)
            .take(10)
            .map(|run| number(run.get("durationMs")))
            .collect(),
        0.5,
    );
    let ratio = latest_ms.zip(baseline).and_then(|(latest, baseline)| {
        if baseline != 0.0 {
            Some(latest / baseline)
        } else {
            None
        }
    });
    json!({
        "latestMs": latest_ms.map(json_number).unwrap_or(Value::Null),
        "baselineMedianMs": baseline.map(json_number).unwrap_or(Value::Null),
        "ratio": ratio.map(json_number).unwrap_or(Value::Null),
        "regression": ratio.is_some_and(|ratio| baseline.unwrap_or(0.0) >= 500.0 && ratio >= 1.2),
    })
}

pub(crate) fn run_history_value(
    harness: &Path,
    app_id: &str,
    target: Option<&str>,
    limit: usize,
) -> Result<Value, Failure> {
    let mut root = harness.join("test-results").join(app_id);
    if let Some(target) = target {
        root.push(target.replace(':', "-"));
    }
    let mut runs = manifests_below(&root)?
        .iter()
        .filter_map(|path| run_record(path))
        .filter(|run| {
            target.is_none_or(|target| run.get("target").and_then(Value::as_str) == Some(target))
        })
        .collect::<Vec<_>>();
    runs.sort_by(|left, right| {
        let left = left
            .get("startedAt")
            .map(Value::to_string)
            .unwrap_or_else(|| "undefined".to_string());
        let right = right
            .get("startedAt")
            .map(Value::to_string)
            .unwrap_or_else(|| "undefined".to_string());
        right.cmp(&left)
    });
    runs.truncate(limit.max(1));
    let count_status = |wanted: &str| {
        runs.iter()
            .filter(|run| run.get("status").and_then(Value::as_str) == Some(wanted))
            .count()
    };
    let passed = count_status("passed");
    let failed = count_status("failed");
    let infrastructure_failures = runs
        .iter()
        .filter(|run| {
            run.get("status").and_then(Value::as_str) == Some("failed")
                && run.get("failureClass").and_then(Value::as_str) == Some("infrastructure")
        })
        .count();
    let product_failures = failed - infrastructure_failures;
    let tests = test_history(&runs);
    let performance = performance_trend(&runs);
    Ok(json!({
        "schemaVersion": 2,
        "appId": app_id,
        "target": target,
        "generatedAt": now(),
        "summary": {
            "runs": runs.len(),
            "passed": passed,
            "failed": failed,
            "blocked": count_status("blocked"),
            "canceled": count_status("canceled"),
            "productFailures": product_failures,
            "infrastructureFailures": infrastructure_failures,
            "passRate": if passed + product_failures > 0 {
                json_number(passed as f64 / (passed + product_failures) as f64)
            } else { Value::Null },
            "flakyTests": tests.iter().filter(|test| test.get("flaky").and_then(Value::as_bool) == Some(true)).count(),
            "latestRunId": runs.first().and_then(|run| run.get("runId")).cloned().unwrap_or(Value::Null),
            "lastGreenRunId": runs.iter().find(|run| run.get("status").and_then(Value::as_str) == Some("passed"))
                .and_then(|run| run.get("runId")).cloned().unwrap_or(Value::Null),
            "performanceRegression": performance.get("regression").cloned().unwrap_or(Value::Bool(false)),
        },
        "performance": performance,
        "journeys": journey_history(&runs),
        "tests": tests,
        "runs": runs,
    }))
}

pub fn history(harness: &Path, app_id: &str, target: Option<&str>, limit: usize) -> Answer {
    print_json(&run_history_value(harness, app_id, target, limit)?)
}
